File: desktop/src/features/onboarding/authCrypto.ts
/**
 * Client-side derivation for email and password accounts.
 *
 * The password never leaves this computer. What goes to the relay is
 * `deriveAuthKey`'s output, which proves the password is known without
 * revealing it, and which is a different value from the one that unlocks the
 * saved account, so the relay learns nothing that helps it open the account.
 */

/** Crockford base32: no I, L, O or U, so a code cannot be misread. */
export const CROCKFORD_ALPHABET = "0123456789ABCDEFGHJKMNPQRSTVWXYZ";

const PBKDF2_ITERATIONS = 600_000;
const GROUP_LEN = 5;
const GROUP_COUNT = 4;

/** Canonical form of an address: trimmed and lowercased. */
export function normaliseEmail(raw: string): string {
  return raw.trim().toLowerCase();
}

function toHex(buffer: ArrayBuffer): string {
  return Array.from(new Uint8Array(buffer))
    .map((byte) => byte.toString(16).padStart(2, "0"))
    .join("");
}

async function sha256(text: string): Promise<ArrayBuffer> {
  return crypto.subtle.digest("SHA-256", new TextEncoder().encode(text));
}

/**
 * Canonical form of a password: NFKC, the same normalisation NIP-49 applies
 * before its scrypt.
 *
 * Without this, a composed and a decomposed spelling of the same accented
 * password derive different keys, so which one works depends on the keyboard
 * that typed it. iOS and macOS routinely produce decomposed text where other
 * sources produce composed.
 */
export function normalisePassword(password: string): string {
  return password.normalize("NFKC");
}

async function pbkdf2AuthKey(email: string, password: string): Promise<string> {
  const salt = await sha256(`colony-auth-v1:${normaliseEmail(email)}`);
  const material = await crypto.subtle.importKey(
    "raw",
    new TextEncoder().encode(password),
    "PBKDF2",
    false,
    ["deriveBits"],
  );
  const bits = await crypto.subtle.deriveBits(
    { name: "PBKDF2", salt, iterations: PBKDF2_ITERATIONS, hash: "SHA-256" },
    material,
    256,
  );
  return toHex(bits);
}

/**
 * Derive the value the relay checks against its stored hash.
 *
 * The salt comes from the address rather than the server, so a second computer
 * can derive this from the password alone with no round trip before the user
 * has typed anything.
 *
 * The password is normalised first, so the same typed password derives the
 * same key whatever keyboard produced it. Accounts created before that was
 * true are reached through {@link deriveLegacyAuthKey} instead.
 */
export async function deriveAuthKey(
  email: string,
  password: string,
): Promise<string> {
  return pbkdf2AuthKey(email, normalisePassword(password));
}

/**
 * Derive the key an account created before signup normalised would have.
 *
 * Only for the sign-in fallback. Normalising at signup fixes new accounts, but
 * an account created from a decomposed password already has an `auth_hash`
 * over those exact bytes, and normalising everywhere would lock its owner out
 * of an account that works today. Sign-in therefore tries the normalised key
 * and falls back to this one.
 *
 * Removable once v2 re-keying lands; see the mobile-account-signin epic
 * artifact for what that takes.
 */
export async function deriveLegacyAuthKey(
  email: string,
  password: string,
): Promise<string> {
  return pbkdf2AuthKey(email, password);
}

/** Generate a recovery code in `XXXXX-XXXXX-XXXXX-XXXXX` form. */
export function generateRecoveryCode(): string {
  const bytes = new Uint8Array(GROUP_LEN * GROUP_COUNT);
  crypto.getRandomValues(bytes);
  const groups: string[] = [];
  for (let index = 0; index < GROUP_COUNT; index += 1) {
    let group = "";
    for (let offset = 0; offset < GROUP_LEN; offset += 1) {
      // The alphabet is exactly 32 characters, so a 5-bit slice is unbiased.
      group +=
        CROCKFORD_ALPHABET[bytes[index * GROUP_LEN + offset] & 0b0001_1111];
    }
    groups.push(group);
  }
  return groups.join("-");
}

/** Lowercase hex SHA-256 of a recovery code, after the same normalisation the relay applies. */
export async function hashRecoveryCode(code: string): Promise<string> {
  return toHex(await sha256(code.trim().toUpperCase()));
}

File: mobile/lib/shared/auth/auth_crypto.dart
/// Client-side derivation for email and password accounts.
///
/// The mirror of `desktop/src/features/onboarding/authCrypto.ts`. The password
/// never leaves the device: what reaches the relay is [deriveAuthKey]'s output,
/// which proves the password is known without revealing it.
///
/// Every constant here is a wire format shared with desktop and with the
/// relay's `is_lowercase_hex(&request.auth_key, 64)` check. A derivation that
/// differs from desktop's by one byte reports `invalid_credentials` for a
/// correct password, with nothing in any log to say why, so the values are
/// pinned by a shared test vector rather than trusted.
library;

import 'dart:convert';
import 'dart:typed_data';

import 'package:pointycastle/export.dart';
import 'package:unorm_dart/unorm_dart.dart' as unorm;

import 'password_kdf.dart';

/// Iterations desktop uses. `authCrypto.ts` calls this PBKDF2_ITERATIONS.
const authKeyIterations = 600000;

/// Bytes of derived key, rendered as 64 hex characters.
const authKeyLength = 32;

/// Salt prefix. Deriving the salt from the address rather than the server lets
/// a second device derive this from the password alone, with no round trip.
const authKeySaltPrefix = 'colony-auth-v1:';

/// A derived auth key and the implementation that produced it.
class AuthKeyDerivation {
  const AuthKeyDerivation({required this.authKey, required this.servedBy});

  /// 64 lowercase hex characters, the shape the relay validates.
  final String authKey;

  final PasswordKdfBackend servedBy;
}

/// Canonical form of an address: trimmed and lowercased.
///
/// Plus-addressing is preserved deliberately, matching `normalise_email` in
/// `crates/buzz-auth/src/account_crypto.rs`: a tagged address is a separate
/// account because a user who tags one expects it to stay separate.
String normaliseEmail(String raw) => raw.trim().toLowerCase();

/// The 32 raw salt bytes for an address.
///
/// Raw digest bytes, not their hex rendering. Desktop passes the `ArrayBuffer`
/// from `crypto.subtle.digest` straight into PBKDF2, and hex here would double
/// the salt's length and change every key.
Uint8List authKeySalt(String email) {
  final digest = SHA256Digest();
  final input = utf8.encode('$authKeySaltPrefix${normaliseEmail(email)}');
  return digest.process(Uint8List.fromList(input));
}

/// Canonical form of a password: NFKC, the same normalisation NIP-49 applies
/// before its scrypt.
///
/// Which spelling a keyboard emits is not something a user chooses. iOS and
/// macOS routinely produce decomposed text where other sources produce
/// composed, and without this the same typed password derives a different key
/// depending on where it was typed.
String normalisePassword(String password) => unorm.nfkc(password);

/// Derive the value the relay checks against its stored hash.
///
/// The password is normalised first, matching desktop's `deriveAuthKey` and
/// NIP-49's own handling. Accounts created before that was true are reached
/// through [deriveLegacyAuthKey] instead.
Future<AuthKeyDerivation> deriveAuthKey({
  required String email,
  required String password,
  PasswordKdf kdf = const PlatformPasswordKdf(),
  bool normalise = true,
}) async {
  final result = await kdf.pbkdf2HmacSha256(
    password: Uint8List.fromList(
      utf8.encode(normalise ? normalisePassword(password) : password),
    ),
    salt: authKeySalt(email),
    iterations: authKeyIterations,
    length: authKeyLength,
  );
  return AuthKeyDerivation(
    authKey: _toHex(result.key),
    servedBy: result.servedBy,
  );
}

String _toHex(Uint8List bytes) {
  final buffer = StringBuffer();
  for (final byte in bytes) {
    buffer.write(byte.toRadixString(16).padLeft(2, '0'));
  }
  return buffer.toString();
}

/// Derive the key an account created before signup normalised would have.
///
/// Only for the sign-in fallback. Normalising at signup fixes new accounts,
/// but an account created from a decomposed password already has an
/// `auth_hash` over those exact bytes, and normalising everywhere would lock
/// its owner out of an account that works today.
///
/// Removable once v2 re-keying lands; see the `account-key-normalisation-v2`
/// epic artifact for what that takes.
Future<AuthKeyDerivation> deriveLegacyAuthKey({
  required String email,
  required String password,
  PasswordKdf kdf = const PlatformPasswordKdf(),
}) =>
    deriveAuthKey(email: email, password: password, kdf: kdf, normalise: false);
